test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,199 @@
+func (ex *examplePlugin) getJsonFilePath(claimUID string) string {
+	return filepath.Join(ex.config.cdiDir, ex.config.driverName+"-"+claimUID+".json")
+}
+
+// isKeyValid returns true if the key is valid.
+func (ex *examplePlugin) isKeyValid(key string) bool {
+	for _, k := range ex.resourceHandleKeys {
+		if k == key {
+			return true
+		}
+	}
+	return false
+}
+
+// validateResourceHandle checks if the resourceHandle has only the valid keys that driver writer decided.
+func (ex examplePlugin) validateResourceHandle(rh map[string]string) error {
+	for key := range rh {
+		if !ex.isKeyValid(key) {
+			return fmt.Errorf("invalid key for ResourceHandle")
+		}
+	}
+	return nil
+}
+
+// startPlugin starts servers that are necessary for plugins and registers the plugin with kubelet.
+func (ex examplePlugin) startPlugin() error {
+	klog.Infof("Starting kubelet plugin")
+
+	// create CDI directory if not exists
+	if err := os.MkdirAll(ex.config.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("failed to create directory: %v", err)
+	}
+
+	// create DRA driver socket directory if not exists.
+	if err := os.MkdirAll(filepath.Dir(ex.config.draAddress), 0750); err != nil {
+		return fmt.Errorf("failed to create DRA driver socket directory: %v", err)
+	}
+
+	// Run a grpc server for the driver, that listens on draAddress
+	if err := plugin.StartNonblockingGrpcServer(ex.config.draAddress, &ex); err != nil {
+		return fmt.Errorf("failed to start a nonblocking grpc server: %v", err)
+	}
+
+	// Run a registration server and registers plugin with kubelet
+	if err := plugin.StartRegistrar(ex.config.driverName, ex.config.draAddress, ex.config.pluginRegistrationPath); err != nil {
+		return fmt.Errorf("failed to start registrar: %v", err)
+	}
+
+	return nil
+
+}
+
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+
+	const (
+		deviceName = "exampledevice"
+		cdiVersion = "0.2.0"
+	)
+	kind := ex.config.driverName + "/" + deviceName
+
+	klog.Infof("NodePrepareResource is called: request: %+v", req)
+	// create CDI Files
+
+	// create spec
+	klog.Infof("Creating CDI Files")
+
+	// resourceHandle in the form of map[string]string
+	var decodedResourceHandle map[string]string
+	if err := json.Unmarshal([]byte(req.ResourceHandle), &decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to unmarshal resourceHandle: %v", err)
+	}
+	// validate if resourceHandle
+	if err := ex.validateResourceHandle(decodedResourceHandle); err != nil {
+		return nil, fmt.Errorf("failed to validate resourceHandle: %v", err)
+	}
+
+	containerEditsInDevices := (&cdiapi.ContainerEdits{}).Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{}})
+	for key, val := range decodedResourceHandle {
+		containerEditsInDevices.Append(&cdiapi.ContainerEdits{ContainerEdits: &specs.ContainerEdits{
```

Comment:
this looks too heavy to me. What about just preparing []string  and then assign it to Env?

Reply:
Why []string? IMO, since yaml has key/value fair, it does not make sense to make it []string.

Reply:
And,, actually how []string?
[]string{'some', 'a', 'another', 'b'}
How can you put them just to Env? Env should have []string{some=a, another=b}

Reply:
Env has []string type as you've mentioned above. That's what I meant - instead of using those heavy things you can just make a []string variable, prepare its content and then assign it to the spec field.

Reply:
I know Env has []string, my point is 'what is the way to make string to []string'? Unmarshal does not work.!
`package main

import (
	"encoding/json"
	"fmt"
)

func main() {
	jsonStr := "{\"user_a\":\"b\",\"user_c\":\"c\"}"
	var unmarshaled []string
	if err := json.Unmarshal([]byte(jsonStr), &unmarshaled); err != nil {
		fmt.Println(err)
	}
	fmt.Println(unmarshaled)

}
---
json: cannot unmarshal object into Go value of type []string
[]

Program exited.
`

Reply:
how about this? https://go.dev/play/p/SNiXg4yxWBK

Reply:
Slight nit - this is impler:
`var unmarshaled map[string]string`

Reply:
Aaa,, I think I misunderstood what you mean. You were pointing out the 'Append'ing many times part. Yeah, I agree that it was stupid idea to do. I should make []string first.